management-node/server.go

```diff
@@ -0,0 +1,530 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/bgzzz/go-schedule/common"
+	pb "github.com/bgzzz/go-schedule/proto"
+	"github.com/bgzzz/go-schedule/wrpc"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+// ManagementNode main structure holding
+// management node grpc server logic
+type ManagementNode struct {
+	// workerNodePool is sync map of WorkerNodes
+	// that are active and connected to management
+	// node
+	workerNodePool    map[string]*wrpc.WorkerRPCClient
+	workerNodePoolMtx sync.RWMutex
+
+	// etcd is link to db
+	etcd *clientv3.Client
+
+	// scheduledTasks is sync map of tasks that are
+	// scheduled at the moment
+	scheduledTasks    map[string]*Task
+	scheduledTasksMtx sync.RWMutex
+}
+
+// Etcd prefixes that are used for storing the objects to
+// etcd
+const (
+	EtcdWorkerPrefix = "worker_"
+	EtcdTaskPrefix   = "task_"
+)
+
+// AddWorkerNode thread safely adds worker node to worker pool and etcd
+// errors in case of data inconsistency
+func (mn *ManagementNode) AddWorkerNode(wn *wrpc.WorkerRPCClient) error {
+	mn.workerNodePoolMtx.Lock()
+	defer mn.workerNodePoolMtx.Unlock()
+
+	if _, ok := mn.workerNodePool[wn.WN.Id]; ok {
+		return fmt.Errorf("There is a worker node with the same id %s in the worker node pool", wn.WN.Id)
+	}
+
+	log.Debugf("Add worker node %s to the worker node pool", wn.WN.Id)
+	mn.workerNodePool[wn.WN.Id] = wn
+
+	wn.WN.State = common.WorkerNodeStateConnected
+
+	err := mn.SetToDb(wn, EtcdWorkerPrefix+wn.WN.Id)
+	if err != nil {
+		log.Error(err.Error())
```

Comment:
As a general rule of thumb, all errors should be propagated to the top-level caller and logged there once, instead of logging each error individually.

```go
if err != nil {
  return trace.Wrap(err)
}
```

Reply:
done